File: internal/paths/paths.go
package paths

import (
	"os"
	"path/filepath"
)

const SwarfDirName = ".swarf"

var (
	ConfigDir        = xdg("XDG_CONFIG_HOME", ".config") + "/swarf"
	StoreDir         = xdg("XDG_DATA_HOME", ".local/share") + "/swarf"
	GlobalConfigTOML = ConfigDir + "/config.toml"
	DrawersTOML      = ConfigDir + "/drawers.toml"
	PIDFile          = ConfigDir + "/daemon.pid"
	LogFile          = ConfigDir + "/daemon.log"
)

func xdg(env, fallback string) string {
	if v := os.Getenv(env); v != "" {
		return v
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, fallback)
}

func SwarfDir(hostRoot string) string     { return filepath.Join(hostRoot, SwarfDirName) }
func LinksDir(hostRoot string) string     { return filepath.Join(SwarfDir(hostRoot), "links") }
func StoreProjectDir(hostRoot string) string { return filepath.Join(StoreDir, ProjectSlug(hostRoot)) }

func ProjectSlug(hostRoot string) string {
	resolved, err := filepath.EvalSymlinks(hostRoot)
	if err != nil {
		resolved = hostRoot
	}
	abs, err := filepath.Abs(resolved)
	if err != nil {
		return filepath.Base(resolved)
	}
	return filepath.Base(abs)
}

// FindHostRoot walks up from start looking for a .swarf/ directory or symlink.
func FindHostRoot(start string) string {
	if start == "" {
		start, _ = os.Getwd()
	}
	current, _ := filepath.Abs(start)
	for {
		candidate := filepath.Join(current, SwarfDirName)
		fi, err := os.Lstat(candidate)
		if err == nil && (fi.IsDir() || fi.Mode()&os.ModeSymlink != 0) {
			return current
		}
		parent := filepath.Dir(current)
		if parent == current {
			return ""
		}
		current = parent
	}
}

// IsDir checks if path exists and is a directory.
func IsDir(path string) bool {
	fi, err := os.Stat(path)
	return err == nil && fi.IsDir()
}

File: src/swarf/paths.py
"""Path constants and helpers for swarf."""

from __future__ import annotations

import os
from pathlib import Path

SWARF_DIR_NAME = ".swarf"

# XDG-compliant paths
CONFIG_DIR = Path(os.environ.get("XDG_CONFIG_HOME", "~/.config")).expanduser() / "swarf"
STORE_DIR = Path(os.environ.get("XDG_DATA_HOME", "~/.local/share")).expanduser() / "swarf"
GLOBAL_CONFIG_TOML = CONFIG_DIR / "config.toml"
DRAWERS_TOML = CONFIG_DIR / "drawers.toml"
PID_FILE = CONFIG_DIR / "daemon.pid"
LOG_FILE = CONFIG_DIR / "daemon.log"


def swarf_dir(host_root: Path) -> Path:
    """Return the .swarf directory for a given host root."""
    return host_root / SWARF_DIR_NAME


def links_dir(host_root: Path) -> Path:
    """Return the path to the links directory."""
    return swarf_dir(host_root) / "links"


def project_slug(host_root: Path) -> str:
    """Derive a project slug from the host repo root directory name."""
    return host_root.resolve().name


def store_project_dir(host_root: Path) -> Path:
    """Return the project's directory inside the central store."""
    return STORE_DIR / project_slug(host_root)


def find_host_root(start: Path | None = None) -> Path | None:
    """Walk up from start looking for a .swarf/ directory or symlink.

    Returns the parent of .swarf/ (the host root), or None if not found.
    """
    current = (start or Path.cwd()).resolve()
    while True:
        candidate = current / SWARF_DIR_NAME
        if candidate.is_dir() or candidate.is_symlink():
            return current
        parent = current.parent
        if parent == current:
            return None
        current = parent
